designs/extensible-cli-and-scaffolding-plugins-phase-2.md

```diff
@@ -0,0 +1,203 @@
+    boilerplate string
+
+    // resource contains the information of the API that is being scaffolded.
+    resource *resource.Resource
+}
+```
+
+One of the plugin hooks allows plugins to set up their own flags. Commong flags (like `--group`, `--version`, `--kind`) are defined by the plugin system itself, and the resource created with those values populated will be passed to the plugins. `create api` and `create webhook` commands would require the GVK being passed as well to the plugins. Currently, `InjectResource` is a required hook for `create api` and `create webhook` commands and the entire resource (which contains the group, version and kind) is being passed to the plugins for these commands.
+
+All the raw flags of the plugins would have to be passed down to the plugins through injector. The plugin specific flags would have to be specified in the project configuration by the user. And the current implmentation is such that, every subcommand can optionally implement the `InjectConfig` hook which passes the project configuration to the plugins and also enable the plugins to modify it if required.
+
+Along with flags, boilerplate with empty copyright or an apache license and in go-style comments is also being passed down to the plugins through injector. If other kind of comments were to be supported, this logic would have to be added (scaffold markers do support YAML and Go-style comments).
+
+Additionally, we would pass the entire CLI environment to the plugin, as anything executing in subshell environment would be able to read the environment variables even if it's being restricted in terms of what Kubebuilder passes to it.
+
+For displaying help text information of the plugins, `UpdateMetadata` hook is the one that allows to modify the description and examples for the `--help` message of plugins. Currently, there is no support for help implementaion of plugins so logic would have to be added to the `UpdateMetadata` hook to display plugin specific help.
+
+#### Handling plugin failures across the chain
+
+If any plugin in the chain fails, we will halt the chain execution and report errors as one plugin may be dependent on the success of another. We can use `stderr` for the plugin to write back warnings/errors in a structured format that Kubebuilder understands and return a process exit code 1. The successful return type with exit code 0 contains the list of files that were scaffolded and the file contents will be given back to Kubebuilder.
+
+In the case of plugin failures even if one plugin across the chain fails, all the files that were scaffolded already until that point will not be written to disk to prevent a half committed state.
+
+#### Open questions
+
+* Do we want to support the addition of new arbitrary subcommands other than the subcommands (init, create api, create webhook) that we already support?
+
+* Do we need to discover flags by calling the plugin binary or should we have users define them in the project configuration?
```

Comment:
The PROJECT is the config of the project scaffold and not of the CLI tool. 
I understand that we need to inform the tool/CLI where the plugins are for it to be able to use them. Then, the data scaffold in the PROJECT file should respect and work as it works for the plugins which are shipped in tool binary.

For example, if we run `create api` with the external plugin we will be able to know that by looking the PROJECT file such as we can check that by looking the plugins attr.  https://github.com/operator-framework/operator-sdk/blob/master/testdata/go/v3/memcached-operator/PROJECT#L4-L6

Reply:
@rashmigottipati I believe we should discover flags by calling the plugin binary.

Reply:
My understanding was that we do need to discover flags. However, my initial idea was to use project configuration for that where users can put in supported flags for each plugin, which is probably not ideal. 

@estroz mentioned that we don't necessarily have to any flag discovery and we could just pass the whole command line and pass it directly to the plugin binary.